opal/mca/accelerator/cuda/accelerator_cuda_component.c

```diff
@@ -0,0 +1,333 @@
+     * paths from the system.  For the second loop, set /usr/lib64 to
+     * the search path and try again.  This is done to handle the case
+     * where we have both 32 and 64 bit libcuda.so libraries
+     * installed.  Even when running in 64-bit mode, the /usr/lib
+     * directory is searched first and we may find a 32-bit
+     * libcuda.so.1 library.  Loading of this library will fail as the
+     * OPAL DL framework does not handle having the wrong ABI in the
+     * search path (unlike ld or ld.so).  Note that we only set this
+     * search path after the original search.  This is so that
+     * LD_LIBRARY_PATH and run path settings are respected.  Setting
+     * this search path overrides them (rather then being
+     * appended). */
+
+    j = 0;
+    while (searchpaths[j] != NULL) {
+        i = 0;
+        while (cudalibs[i] != NULL) {
+            char *filename = NULL;
+            char *str = NULL;
+
+            /* If there's a non-empty search path, prepend it
+             * to the library filename */
+            if (strlen(searchpaths[j]) > 0) {
+                opal_asprintf(&filename, "%s/%s", searchpaths[j], cudalibs[i]);
+            } else {
+                filename = strdup(cudalibs[i]);
+            }
+            if (NULL == filename) {
+                opal_show_help("help-accelerator-cuda.txt", "No memory", true,
+                               OPAL_PROC_MY_HOSTNAME);
+                return NULL;
+            }
+
+            retval = opal_dl_open(filename, false, false, &libcuda_handle, &str);
+            if (OPAL_SUCCESS != retval || NULL == libcuda_handle) {
+                if (NULL != str) {
+                    opal_argv_append(&errsize, &errmsgs, str);
+                } else {
+                    opal_argv_append(&errsize, &errmsgs, "opal_dl_open() returned NULL.");
+                }
+                opal_output_verbose(10, opal_accelerator_base_framework.framework_output, "CUDA: Library open error: %s",
+                                    errmsgs[errsize - 1]);
+            } else {
+                opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
+                                    "CUDA: Library successfully opened %s", cudalibs[i]);
+                found_libraries = true;
+                break;
+            }
+            i++;
+            free(filename);
+        }
+        if (true == found_libraries) {
+            break; /* Break out of outer loop */
+        }
+        j++;
+    }
+
+    if (true != found_libraries) {
+        errmsg = opal_argv_join(errmsgs, '\n');
+        if (opal_warn_on_missing_libcuda) {
+            opal_show_help("help-accelerator-cuda.txt", "dlopen failed", true, errmsg);
+        }
+    }
+    opal_argv_free(errmsgs);
+    free(errmsg);
+
+    if (true != found_libraries) {
+        return NULL;
+    }
+
+    if (OPAL_SUCCESS != accelerator_cuda_populate_func_table(libcuda_handle)) {
```

Comment:
I noticed that the ROCM init routine acquires a lock around a similar set of code in its init routine.   Is one needed here?   (Should the lock in the ROCM init routine be removed @edgargabriel ?)

Reply:
The accelerator init functions will all be during opal_init, which is serialized.  So I don't think a lock is needed here in this case.  It would be needed in the lazy init case that used to exist when this code was in opal/cuda.

Reply:
I got rid of the locking in the init routine